### DISCUSSION TITLE
How to delete a store value by key?

### DISCUSSION THREAD

**Discussion by benjaminkarat** [ai/nanostores#157] [Q&A]
I am trying to create a cart that has the option to remove items from it. I didn't find a way  to delete values of a store so far, is that possible? If I manually delete a item from the localStorage it seems to not sync anymore

**Comment by ai** [comment]
Use value encoder and store an array in the `persistentAtom`

https://github.com/nanostores/persistent#value-encoding

```js
import { persistentAtom } from '@nanostores/persistent'

export const cart = persistentAtom('draft', [], {
  encode (value) {
    return JSON.stringify(value)
  },
  decode (value ) {
    try {
      return JSON.parse(value)
    } catch() {
      return value
    }
  }
})

function addToCart (item) {
  cart.set([...cart.get(), item])
}

function removeFromCart (id) {
  cart.set(cart.get().filter(i => i.id !== id))
}
```

**Reply by benjaminkarat** [reply]
Thanks @ai based on your answer I came up with a way to delete a value from a persistentMap. It deletes the item as I can see in the localStorage, strangely it doesn't sync with the store that's saved in Vue and after reloading the page it's in the localStorage again, is it possible that it's due to the way I initialize it in Vue:

Initializing the cartItems (Vue):
```Javascript
setup () {
      let isCartOpen = useStore(isCartOpen_)
      let cartItems = useStore(cartItems_)
      // console.log('cartItems SETUP', cartItems, typeof cartItems)
      return { isCartOpen, cartItems }
    }
```

Code to remove cart item:
```Javascript
export function removeCartItem(id) {
  // filter removed item from cart and set new cart
  const newCart = Object.keys(cartItems_.get()).reduce((acc, key) => {
    if (key !== id) {
      acc[key] = cartItems_.get()[key]
    }
    return acc
  }
  , {})
  cartItems_.set(newCart)
  console.log('newCart', newCart)
}
```
